### ISSUE TITLE
[FEAT] Resolve detection class head reuse by class name instead of class count

### ISSUE THREAD

**Issue by liopeer** [lightly-ai/lightly-train#985]
### 💡 Is your feature request related to a problem?

#982 fixed the case where the checkpoint and the new model have a *different* number of classes. It explicitly kept the old behavior when the counts match, and #983 implements exactly that.

That leaves the equal-count case, where `_reuse_or_reinit` in `src/lightly_train/_task_models/object_detection_components/hooks.py` still maps checkpoint class row `k` onto new-dataset class `k` purely because the two counts are equal:

```python
num_classes_checkpoint = checkpoint_weight.shape[0]
num_classes_module = getattr(head_module, "out_features", None)
if num_classes_module is None or num_classes_checkpoint == num_classes_module:
    return False  # reuse the checkpoint rows unchanged
```

Fine-tuning a 3-class traffic checkpoint (`{0: "car", 1: "bus", 2: "truck"}`) onto a 3-class medical dataset (`{0: "polyp", 1: "ulcer", 2: "bleeding"}`) therefore warm-starts the classification head from rows with unrelated semantics, with no warning. The class count is a proxy for "same label space" that does not hold in general, and equal counts are common for small custom datasets.

The same applies to `denoising_class_embed_reuse_or_reinit_hook`, and to LTDETRv2 instance segmentation, since `ECSegTransformer` subclasses `DFINETransformer` (`instance_segmentation_components/edgecrafter_decoder.py`) and inherits both hook registrations.

### 🧰 Describe the solution you'd like

Resolve reuse per class name instead of per index:

- Name present in both → reuse the checkpoint row, remapped to the new index.
- Name missing from the checkpoint → keep the module's initialized row.
- Log how many rows were reused vs re-initialized.

This makes the equal-count case correct and also recovers partial reuse when the counts differ, e.g. fine-tuning COCO onto `{0: "person", 1: "car"}` would keep those two pretrained rows instead of discarding everything.

The class names are available in the checkpoint under `model_init_args["classes"]`, and the new names on `LTDETRObjectDetection.classes` (`ltdetr_object_detection/task_model.py`).

The main work is plumbing rather than the matching itself. The hooks are registered on the decoder (`dfine_decoder.py`, `rtdetrv2_decoder.py`), and `task_model.py` passes the decoder only `num_classes=len(self.classes)`, so the decoder currently has no access to either set of names. Options:

- Pass the class names into the decoder so the existing hooks can use them.
- Do the remapping at the task-model level and leave the hooks handling only the shape mismatch.

### 🛠 Alternatives you've considered

- **Keep count-based matching, warn when counts match.** Cheapest, no behavior change, no plumbing, but leaves the wrong-semantics warm start as the default and pushes the decision onto the user.
- **Always re-initialize.** Removes the failure mode but throws away the genuine warm start when the label space really is the same, which is the common case when resuming or iterating on one dataset.

### 📝 Additional context

Follow-up to #982 / #983. `eomt/hooks.py::class_head_reuse_or_reinit_hook` and `picodet_object_detection/pico_head.py::picodet_gfl_cls_reuse_or_reinit_hook` decide the same way, so whatever shape this takes should probably cover them too.
